### PR TITLE
cpu/atmega_common: adapt to new i2c api

### PR DESCRIPTION
**Pull request by mali**
### Contribution description

Rework atmega i2c to match new i2c interface.
Only i2c_read_bytes and i2c_write_bytes are implemented.

WIP : Comments welcome !

### Issues/PRs references

#6577

### CODE REVIEW

**Comment by ZetaR60** [comment]
@mali I have gone over this PR and #9165 several times, but I have been unable to track down the problem I have been having in #9165. The slave does not ack on the bus with the new code, but I have confirmed that the old code works fine. The two code sets don't have a lot of differences in behavior prior to the failure, so it is probably just something I have overlooked. Could you take a look and see if you spot anything that could be causing it?

If anyone has any other I2C devices that have been ported and can be tested with this PR it would be much appreciated!

File: cpu/atmega_common/periph/i2c.c [hunk 1]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+        TWCR = (1 << TWEN) | (1 << TWINT);
+        DEBUG("Wait for byte\n");
+
         /* Wait for TWINT Flag set. This indicates that DATA has been received.*/
         while (!(TWCR & (1 << TWINT))) {}
+
         /* receive data byte */
-        my_data[i] = TWDR;
-        DEBUG("Byte %i received\n", i+1);
+        my_data[0] = TWDR;
```
Comment (aabadie) [inline cpu/atmega_common/periph/i2c.c]: This looks like an introduced bug.

File: cpu/atmega_common/periph/i2c.c [hunk 2]
```diff
@@ -326,37 +265,37 @@ static int _start(uint8_t address, uint8_t rw_flag)
...
 }
 
+/* TODO : const uint8_t data instead of *data */
 static int _write(const uint8_t *data, int length)
 {
     for (int i = 0; i < length; i++) {
         /* Load DATA into TWDR Register.
          * Clear TWINT bit in TWCR to start transmission of data */
-        TWDR = data[i];
+        TWDR = data[0];
```
Comment (aabadie) [inline cpu/atmega_common/periph/i2c.c]: Same here

File: cpu/atmega_common/periph/i2c.c [hunk 3]
```diff
@@ -57,18 +64,24 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
 
     /* check if the line is valid */
     if (dev >= I2C_NUMOF) {
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: prefer the use of `assert(dev < I2C_NUMOF)` instead of `if`

File: cpu/atmega_common/periph/i2c.c [hunk 4]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
-    }
-
-    /* now start a new start condition and receive data */
-    return i2c_read_bytes(dev, address, data, length);
-}
 
-int i2c_write_byte(i2c_t dev, uint8_t address, uint8_t data)
-{
-    return i2c_write_bytes(dev, address, &data, 1);
+    return I2C_ACK;
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: I2C_ACK doesn't exist anymore as we switch to errno return code for the whole API. You can replace I2C_ACK by 0

File: cpu/atmega_common/periph/i2c.c [hunk 5]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
 }
 
-int i2c_write_bytes(i2c_t dev, uint8_t address, const void *data, int length)
+int i2c_write_bytes(i2c_t dev, uint16_t addr, const void *data, size_t len, uint8_t flags)
 {
-    int bytes = 0;
-
-    assert(length > 0);
-
+    assert (len > 0);
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: must be switch to :
```
if(data == NULL || len == 0) {
    return -EINVAL;
}
```

File: cpu/atmega_common/periph/i2c.c [hunk 6]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+    if(~(flags & I2C_NOSTART)) {
+        if (_start(addr, ATMEGA_I2C_FLAG_WRITE) != 0) {
+            return I2C_ADDR_NACK;
+        }
     }
 
     /* send out data bytes */
-    bytes = _write(data, length);
+    if(!_write(data, len))
+        return I2C_DATA_NACK;
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: must be switch to `return -EIO;`

File: cpu/atmega_common/periph/i2c.c [hunk 7]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
 
     /* end transmission */
-    _stop();
+    if(~(flags & I2C_NOSTOP))
+        _stop();
 
-    return bytes;
+    return I2C_ACK;
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: must be switch to `return 0;`

File: cpu/atmega_common/periph/i2c.c [hunk 8]
```diff
@@ -120,169 +133,95 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
-    /* return number of bytes send */
-    return bytes;
+    mutex_unlock(&lock);
+    return 0;
 }
 
-void i2c_poweron(i2c_t dev)
+static void i2c_poweron(i2c_t dev)
 {
     assert(dev < I2C_NUMOF);
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: please unify the use of assert for each dev test against I2C_NUMOF

File: cpu/atmega_common/periph/i2c.c [hunk 9]
```diff
@@ -326,37 +265,37 @@ static int _start(uint8_t address, uint8_t rw_flag)
...
+            return 0;
         }
         else {
-            DEBUG("ACK has been received for BYTE %i\n", i+1);
-        }
+            DEBUG("ACK has been received\n");
+        } 
     }
-
     return length;
```
Comment (dylad) [inline cpu/atmega_common/periph/i2c.c]: I2C API functions does not return the length anymore. This should be replaced by 0

**Comment by dylad** [comment]
@mali Any progress on this ?

**Comment by mali** [comment]
@dylad I'm wating for an i2c shield I've ordered. (the same that aabadie had lent me during the Hackaton).
It should arrive soon, i hope, so i'll be able to test for real.

**Comment by MrKevinWeiss** [comment]
@mali Any update?  I can test here.

**Comment by mali** [comment]
@MrKevinWeiss , please do !

**Comment by aabadie** [comment]
@mali, I allowed myself to update this PR with a few things:
- use assert as requested by @dylad 
- fix a few issues related to coding conventions
- improve debug messages
- adapt I2C bus speed switch in device initialization
- adapt atmega based board configuration

**Comment by mali** [comment]
@aabadie , no problem, you're welcome !
I know time is ticking and I do not have as much as I would have liked

**Comment by Aang-3** [comment]
We have just tested it on an Arduino Mega 2560 board using the driver_hts221' test. It works fine.

**Comment by aabadie** [comment]
Thanks for testing @Aang-3. Just in case, I squashed the branch. @dylad, are you happy with the changes ?

**Comment by aabadie** [comment]
Please wait before merging. With @Aang-3, we realized that this PR is broken in the current state.

**Comment by aabadie** [comment]
@dylad, I fixed the driver (plus a couple of other things), so it can be merged now.

**Comment by MrKevinWeiss** [comment]
I am having some issues testing... I need some time to figure it out.

File: cpu/atmega_common/periph/i2c.c [hunk 10]
```diff
@@ -119,175 +133,122 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+    if (data == NULL || len == 0) {
+        return -EINVAL;
     }
 
     uint8_t *my_data = data;
+
     /* send start condition and slave address */
-    if (_start(address, I2C_FLAG_READ) != 0) {
-        return 0;
+    if (~(flags & I2C_NOSTART)) {
```
Comment (MrKevinWeiss) [inline cpu/atmega_common/periph/i2c.c]: I think you mean ! not ~

File: cpu/atmega_common/periph/i2c.c [hunk 11]
```diff
@@ -119,175 +133,122 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
-    }
-
-    /* start transmission and send slave address */
-    if (_start(address, I2C_FLAG_WRITE) != 0) {
-        return 0;
-    }
-
-    /* send register address and wait for complete transfer to be finished*/
-    if (_write(&reg, 1) != 1) {
+    if (~(flags & I2C_NOSTOP)) {
```
Comment (MrKevinWeiss) [inline cpu/atmega_common/periph/i2c.c]: Same here, ! not ~

File: cpu/atmega_common/periph/i2c.c [hunk 12]
```diff
@@ -119,175 +133,122 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+    }
+    /* Check for wrong arguments given */
+    if (data == NULL || len == 0) {
+        return -EINVAL;
     }
 
     /* start transmission and send slave address */
-    if (_start(address, I2C_FLAG_WRITE) != 0) {
-        return 0;
+    if (~(flags & I2C_NOSTART)) {
```
Comment (MrKevinWeiss) [inline cpu/atmega_common/periph/i2c.c]: Almost there, ! not ~

File: cpu/atmega_common/periph/i2c.c [hunk 13]
```diff
@@ -119,175 +133,122 @@ int i2c_init_master(i2c_t dev, i2c_speed_t speed)
...
+        DEBUG("[i2c] i2c_write_bytes: write failed\n");
+        return -EIO;
+    }
 
     /* end transmission */
-    _stop();
-
-    return bytes;
-}
+    if (~(flags & I2C_NOSTOP)) {
```
Comment (MrKevinWeiss) [inline cpu/atmega_common/periph/i2c.c]: finally, ! not ~

**Comment by aabadie** [comment]
> It seems like the flags are not functioning, this made write reg fail the test

Good catch! And to move fast here, I pushed a fixup commit that addresses your comments. Please test again @MrKevinWeiss !

**Comment by MrKevinWeiss** [comment]
ACK, passing all non-critical tests!

**Comment by aabadie** [comment]
Squashed. We can merge if Travis doesn't report know issues.

**Comment by mali** [comment]
:+1:

**Pull request merged**